10_rest/runnerdave/pkg/puppy/rest.go

```
@@ -0,0 +1,89 @@
+type RestServer struct {
+	DB Storer
+}
+
+// SetupRoutes setups the routes for the server
+func (rs *RestServer) SetupRoutes(db Storer) *chi.Mux {
+	r := chi.NewRouter()
+	rs.DB = db
+
+	r.Route("/api/puppy", func(r chi.Router) {
+		r.Post("/", rs.createPuppy)
+		r.Route("/{id}", func(r chi.Router) {
+			r.Put("/", rs.updatePuppy)
+			r.Delete("/", rs.deletePuppy)
+			r.Get("/", rs.getPuppy)
+		})
+	})
+
+	return r
+}
+
+func (rs *RestServer) getPuppy(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	p, err := rs.DB.ReadPuppy(int16(id))
+	if err != nil {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	bytes, _ := json.Marshal(p)
+	_, _ = w.Write(bytes)
+}
+
+func (rs *RestServer) createPuppy(w http.ResponseWriter, r *http.Request) {
+	p := Puppy{}
+	if err := json.NewDecoder(r.Body).Decode(&p); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	if err := rs.DB.CreatePuppy(p); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusCreated)
+}
+
+func (rs *RestServer) updatePuppy(w http.ResponseWriter, r *http.Request) {
+	p := Puppy{}
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	if err = json.NewDecoder(r.Body).Decode(&p); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	if err = rs.DB.UpdatePuppy(int16(id), &p); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func (rs *RestServer) deletePuppy(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	if err = rs.DB.DeletePuppy(int16(id)); err != nil {
+		w.WriteHeader(http.StatusNoContent)
```

I think this is backwards now isn't it? if `DeletePuppy` returns an error, you want to return an error. If no error, you want No Content.

It is viable to handle the case where deleting a non-existent puppy just returns "sure whatever, let's call that done" because the puppy isn't there at the end, but in that case you should probably differentiate between different errors - for example, if you had a SQL DB store implemented, one error returned here may be a "database unreachable" error. In that case, you should return 500 Internal Error.

All that said, the logic here is still a little off - if there is no error, you're still not returning No Content.

---

Hi Cam thanks a lot for your patience, I did a last minute fix without thinking what I wanted to do, my test was actually wrong, but thanks to the returning of proper status codes I understood what was wrong, it was a real privilege to receive your reviews, hope is ok this time